akka-cluster-sharding/src/main/scala/akka/cluster/sharding/SliceRangeShardAllocationStrategy.scala

```diff
@@ -0,0 +1,264 @@
+            else if (neighborSlices.max < slice)
+              findRegionWithUpperNeighbor(slice, maxShards, currentShardAllocations)
+            else
+              None
+
+          val selectedRegion =
+            alternative match {
+              case Some(alternativeRegion) =>
+                if (neighborShards.size >= currentShardAllocations(alternativeRegion).size)
+                  regionWithNeighbor
+                else
+                  alternativeRegion
+              case None =>
+                regionWithNeighbor
+            }
+          Future.successful(selectedRegion)
+
+        } else {
+          Future.successful(regionWithNeighbor)
+        }
+      case None =>
+        Future.successful(allocateWithoutNeighbor(slice, currentShardAllocations))
+    }
+  }
+
+  private def allocateWithoutNeighbor(
+      slice: Int,
+      currentShardAllocations: Map[ActorRef, IndexedSeq[ShardId]]): ActorRef = {
+
+    // sort regions by member age because if a node is added (random address) we don't want rebalance more than necessary
+    val regionsByMbr = regionEntriesFor(currentShardAllocations).map(entry => entry.member -> entry.region).toMap
+    val regions = regionsByMbr.keysIterator.toIndexedSeq.sorted(Member.ageOrdering).map(regionsByMbr(_))
+
+    // first look for an empty region (e.g. rolling update case)
+    regions.reverse.collectFirst { case region if currentShardAllocations(region).isEmpty => region } match {
+      case Some(emptyRegion) => emptyRegion
+      case None =>
+        val rangeSize = NumberOfSlices / regions.size
+        val i = slice / rangeSize
+
+        if (i < regions.size) {
+          regions(i)
+        } else {
+          // This covers the rounding case for the last region, which we just distribute over all regions.
+          // May also happen if member for that region has been removed, but that should be a rare case.
+          regions(slice % regions.size)
+        }
+    }
+
+  }
+
+  override def rebalance(
+      currentShardAllocations: Map[ActorRef, IndexedSeq[ShardId]],
+      rebalanceInProgress: Set[ShardId]): Future[Set[ShardId]] = {
+    import math.max
+    import math.min
+
+    def limit(numberOfShards: Int): Int =
+      max(1, min((relativeLimit * numberOfShards).toInt, absoluteLimit))
+
+    if (rebalanceInProgress.nonEmpty) {
+      // one rebalance at a time
+      emptyRebalanceResult
+    } else {
+      val sortedRegionEntries = regionEntriesFor(currentShardAllocations).toVector.sorted(ShardSuitabilityOrdering)
+      if (!isAGoodTimeToRebalance(sortedRegionEntries)) {
+        emptyRebalanceResult
+      } else {
+        // this is the number of slices per region that we are aiming for
+        val overfill = 1
+        val targetSize = NumberOfSlices / sortedRegionEntries.size + overfill
+        val selected = Vector.newBuilder[ShardId]
+        // FIXME ShardSuitabilityOrdering isn't used, but it seems better to use most shards first, combine them?
```

Comment:
Missing the handling of leaving regions and app version from there because there if we don't want to use it

Reply:
Lol, that sentence was not parseable, just wanted to note that the ShardSuitabilityOrdering looks at member leaving and app version, so if we don't want to use it we should probably do something around those properties as well.